### PR TITLE
Add stochastic sampling to FlowMatchEulerDiscreteScheduler

### PR DESCRIPTION
**Pull request by apolinario**
# What does this PR do?

This PR adds stochastic sampling to FlowMatchEulerDiscreteScheduler based on https://github.com/Lightricks/LTX-Video/commit/b1aeddd7ccac85e6d1b0d97762610ddb53c1b408 `ltx_video/schedulers/rf.py`, which was added with th release of 0.9.6-distilled. I decoupled the next and current sigma to try to get closer to the `rf.py` implementation of the stochastic sampling, but a second pair of eyes on this would be great.

To try it: 
```py
import torch
from diffusers import LTXVideoTransformer3DModel, FlowMatchEulerDiscreteScheduler, LTXPipeline
from diffusers.utils import export_to_video

transformer = LTXVideoTransformer3DModel.from_pretrained(
    "multimodalart/ltxv-2b-0.9.6-distilled",
    subfolder="transformer",
    torch_dtype=torch.bfloat16,
    variant="bf16"
)

scheduler = FlowMatchEulerDiscreteScheduler.from_pretrained(
    "multimodalart/ltxv-2b-0.9.6-distilled",
    subfolder="scheduler"
)

pipe = LTXPipeline.from_pretrained(
    "Lightricks/LTX-Video-0.9.5",
    transformer=transformer,
    scheduler=scheduler, #add or remove the scheduler to see the difference
    torch_dtype=torch.bfloat16,
)
pipe.to("cuda")

prompt = "A woman eating a burger"
negative_prompt = "worst quality, inconsistent motion, blurry, jittery, distorted"
generator = torch.Generator(device="cuda").manual_seed(42)
video = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    width=1216,
    height=704,
    num_frames=121,
    num_inference_steps=8,
    guidance_scale=1,
    generator=generator
).frames[0]

export_to_video(video, "distilled_scheduler.mp4", fps=24)
```

## Who can review?

@yiyixuxu

### CODE REVIEW

**Comment by apolinario** [comment]
@bot /style

File: src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py [hunk 1]
```diff
@@ -437,13 +443,28 @@ def step(
...
+            sigma = self.sigmas[sigma_idx]
+            sigma_next = self.sigmas[sigma_idx + 1]
+
+            current_sigma = sigma
+            next_sigma = sigma_next
             dt = sigma_next - sigma
 
-        prev_sample = sample + dt * model_output
+        # Determine whether to use stochastic sampling for this step
+        use_stochastic = stochastic_sampling if stochastic_sampling is not None else self.config.stochastic_sampling
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py:460]: I think just have this in config is enough no?

**Review by yiyixuxu — APPROVED** [review]
thanks @apolinario

**Comment by apolinario** [comment]
@bot /style

File: src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py [hunk 2]
```diff
@@ -437,13 +443,25 @@ def step(
             lower_mask = sigmas < per_token_sigmas[None] - 1e-6
             lower_sigmas = lower_mask * sigmas
             lower_sigmas, _ = lower_sigmas.max(dim=0)
-            dt = (per_token_sigmas - lower_sigmas)[..., None]
+
+            current_sigma = per_token_sigmas[..., None]
+            next_sigma = lower_sigmas[..., None]
+            dt = next_sigma - current_sigma  # Equivalent to sigma_next - sigma
```
[the remark this answers is not in the document]
Comment (apolinario) [reply src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py]: good catch!

**Comment by yiyixuxu** [comment]
thanks @apolinario !

**Comment by yiyixuxu** [comment]
@nitinmukesh I think it's probably better in `LTXConditionPipeline`, can you try it out?

**Comment by nitinmukesh** [comment]
@apolinario 

Thank you for adding the sampling.

Please could you share few sample outputs which you created. I am not getting good results so want to compare if something wrong in code.
https://github.com/huggingface/diffusers/issues/11359

**Comment by Ednaordinary** [comment]
> Quick question, Should it be LTXPipeline or LTXConditionPipeline? 0.9.5 support was added in LTXConditionPipeline.

LTXPipeline. The model here is 0.9.6-distilled (the only one that uses the stochastic sampler as of now). "0.9.5" is included because the transformer and scheduler from 0.9.6 are inserted, which is fine because nothing else in the pipeline is different from 0.9.5 and there's currently nothing the lightricks 0.9.6 repos. 0.9.6-distilled is guidance distilled so it does not work in the condition pipeline, while 0.9.6 does

**Comment by Ednaordinary** [comment]
Also, I find increasing the schedulers shift while using distilled helps to boost coherence. This is inline with FastVideo (PCM distillation) which says to set the shift to 17. The 1.0 in https://huggingface.co/multimodalart/ltxv-2b-0.9.6-distilled/blob/main/scheduler/scheduler_config.json doesn't seem like a good default. I'm unsure what it is in the original LTX repo

<details><summary>Some different shifts</summary>
<p>

0.25

https://github.com/user-attachments/assets/6362a108-2912-4502-ba8e-4db6d3316637

0.5

https://github.com/user-attachments/assets/889b6121-4015-406b-ab87-3c82e961a774

1.0

https://github.com/user-attachments/assets/86f72972-c9c9-463f-80d9-1621dbb954f7

2.0


https://github.com/user-attachments/assets/2891c66c-59bf-4710-8f58-2046403e079f

4.0

https://github.com/user-attachments/assets/19f391be-7234-4eac-a501-ae127d42ee1d

8.0

https://github.com/user-attachments/assets/a72d0fe8-cd4a-40bd-bf0d-5f4f33402cd3

16.0


https://github.com/user-attachments/assets/70729950-1075-4e5d-9442-52779ba83742

32

https://github.com/user-attachments/assets/80fa3fd2-b462-4da5-9e3e-b4a70cf3d590

64

https://github.com/user-attachments/assets/8af29384-1897-47b4-a962-9c10cb285ee5


</p>
</details> 

16 seems like a good default

**Comment by nitinmukesh** [comment]
Thank you @Ednaordinary 

The information you provided is very helpful. Getting better results than before.

https://github.com/user-attachments/assets/34b7c750-92eb-4f36-b810-5dee50fa2934

**Comment by nitinmukesh** [comment]
249 frames

https://github.com/user-attachments/assets/66f0085f-7de5-4669-8410-b1f7a9cbf5a1

**Comment by nitinmukesh** [comment]
Sure will try that, thank you. Next gonna try if distilled support image to video (LTXImageToVideoPipeline).
The speed of the models is OoTW, all these generated on 8 GB VRAM + 16 GB RAM and only 3 minutes for 249 frames.

**Comment by nitinmukesh** [comment]
I2V is working good.

https://github.com/user-attachments/assets/6a74d7a6-b2b5-4ae7-9977-50831168bcaa

**Comment by nitinmukesh** [comment]
@Ednaordinary 

What do you suggest for 0.9.6 dev, LTXPipeline or Conditioning pipeline, in case you tried.

**Comment by Ednaordinary** [comment]
0.9.6 should be the condition pipeline I'm pretty sure
